### ISSUE TITLE
Download hls audio. Key problem

### ISSUE THREAD

**Issue by BoikoIlya** [androidx/media#572]
i download m3u8 format audio using this code: 
val cache = SimpleCache(
            File(context.externalCacheDir!!.path,cachedTracksdbFileName),
            NoOpCacheEvictor(),
            StandaloneDatabaseProvider(context)
        )
val cacheDataSourceFactory =  CacheDataSource.Factory()
            .setCache(simpleCache)
            .setUpstreamDataSourceFactory(DefaultHttpDataSource.Factory());
 val hlsDownloader = HlsDownloader(mediaWithStreamKey, cacheDataSourceFactory)
                hlsDownloader.download(listener)

And i check if track cached using this: 
val length = cache.getCachedLength(Uri.parse(data.url).toString(),0,Long.MAX_VALUE)
        val isCached = cache.isCached(Uri.parse(data.url).toString(),0,length)

The key is track uri. But  tracks that came from my api change url sometimes. that is why i get false when try to check if it is cached. Is it possible to solve this?

**Comment by tianyif** [comment]
Hi @BoikoIlya,

For overriding the default behaviour, you can have a customized implementation of [`CacheKeyFactory`](https://github.com/androidx/media/blob/release/libraries/datasource/src/main/java/androidx/media3/datasource/cache/CacheKeyFactory.java), and [inject](https://github.com/androidx/media/blob/release/libraries/datasource/src/main/java/androidx/media3/datasource/cache/CacheDataSource.java#L149) it in `CacheDataSource.Factory`. Could you please try if this can solve your problem?

**Comment by BoikoIlya** [comment]
I have consider this way. 

> Hi @BoikoIlya,
> 
> For overriding the default behaviour, you can have a customized implementation of [`CacheKeyFactory`](https://github.com/androidx/media/blob/release/libraries/datasource/src/main/java/androidx/media3/datasource/cache/CacheKeyFactory.java), and [inject](https://github.com/androidx/media/blob/release/libraries/datasource/src/main/java/androidx/media3/datasource/cache/CacheDataSource.java#L149) it in `CacheDataSource.Factory`. The method to be implemented takes a `DataSpec` to build the cache key. The current code of `HlsDownloader` (a final subclass of `SegmentDownloader`) only takes the `uri` to create the `DataSpec` (see [example](https://github.com/androidx/media/blob/release/libraries/exoplayer_hls/src/main/java/androidx/media3/exoplayer/hls/offline/HlsDownloader.java#L189)), but it should be fine if your key is going to be derived from part of the uri. Could you please see if this can solve your problem?

I have consider this way, but my url completely change.

**Comment by tianyif** [comment]
@BoikoIlya Do you know in your case if there is anything can be used to identify your dataspec other than the uri? There is a [field `key`](https://github.com/androidx/media/blob/release/libraries/datasource/src/main/java/androidx/media3/datasource/cache/CacheKeyFactory.java#L27) that the dataspec can carry with, this requires change in `HlsDownloader` to build dataspec with the `key` though.

**Comment by BoikoIlya** [comment]
> @BoikoIlya Do you know in your case if there is anything can be used to identify your dataspec other than the uri? There is a [field `key`](https://github.com/androidx/media/blob/release/libraries/datasource/src/main/java/androidx/media3/datasource/cache/CacheKeyFactory.java#L27) that the dataspec can carry with, this requires change in `HlsDownloader` to build dataspec with the `key` though.

Is it possible to use media item id as a key?

**Comment by tianyif** [comment]
Hi @BoikoIlya,

> Is it possible to use media item id as a key?

Does it mean that:  You want to create a local HLS playlist that refers the downloaded media playlists and segments via their local uris, and then use the `MediaItem.mediaId` as the key to look up the url of the created local HLS playlist? When the player is going to play that media item with a new url, then if any local HLS playlist is associated with the same id, then just play with that local playlist instead?

**Comment by BoikoIlya** [comment]
> Hi @BoikoIlya,
> 
> > Is it possible to use media item id as a key?
> 
> Does it mean that: You want to create a local HLS playlist that refers the downloaded media playlists and segments via their local uris, and then use the `MediaItem.mediaId` as the key to look up the url of the created local HLS playlist? When the player is going to play that media item with a new url, then if any local HLS playlist is associated with the same id, then just play with that local playlist instead?

Yes, if i understood in right way. Because media Id is single permanent identifier of audio in my case.

**Comment by BoikoIlya** [comment]
> Hi @BoikoIlya,
> 
> > Is it possible to use media item id as a key?
> 
> Does it mean that: You want to create a local HLS playlist that refers the downloaded media playlists and segments via their local uris, and then use the `MediaItem.mediaId` as the key to look up the url of the created local HLS playlist? When the player is going to play that media item with a new url, then if any local HLS playlist is associated with the same id, then just play with that local playlist instead?

i am sorry, but i am still waiting for reply

**Comment by tianyif** [comment]
@BoikoIlya, we have some requests from other issues to allow guaranteed plumbing from `MediaItem` to `DataSpec`, so this could probably be a near-future enhancement we could add. 

However for your case, I feel plumbing to `DataSpec` doesn't help too much. For example, for a media segment, you cannot use the uri in the `DataSpec` and the plumbed `mediaId` to get the uri of the downloaded segment, so that's why I asked you if you plan to create a local HLS playlist that refers the downloaded media playlists and segments via their local uris, and then use the MediaItem.mediaId as the key to look up the url of the created local HLS playlist. But in this way, this lookup doesn't have to be in the `DataSpec`, then you can have mapping logic in your app to get the url of the created local multi variant playlist with the `mediaId`.

**Comment by tianyif** [comment]
I'm closing this issue for now. Please feel free to file a new issue for more questions related.
